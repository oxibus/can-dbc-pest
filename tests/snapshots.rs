#![cfg(feature = "encoding")]

use std::borrow::Cow;
use std::fs;
use std::path::Path;

use can_dbc_pest::{decode_cp1252, DbcParser, Rule};
use insta::{assert_debug_snapshot, with_settings};
use pest::Parser;

/// Test parsing all DBC files
#[test]
fn test_cantools_dbc_files() {
    let test_dirs = [
        (
            "tests/shared-test-files/dbc-cantools",
            "dbc-cantools",
            true,
            true,
        ),
        ("tests/shared-test-files/canpy", "canpy", true, true),
        ("tests/opendbc/opendbc/dbc", "opendbc", false, false),
    ];
    for (path, file_name, snapshot, cp1251) in test_dirs {
        let path = Path::new(path);
        let snapshot_path = String::from("snapshots-") + file_name;
        // snapshots should go to /tests/snapshots-* directory
        with_settings! {
            { omit_expression => true,
              snapshot_path => snapshot_path ,
              prepend_module_to_snapshot => false },
            {
                test_dbc_files(path, snapshot, cp1251);
            }
        }
    }
}

/// Test parsing all DBC files in the given directory.
fn test_dbc_files(dir: impl AsRef<Path>, snapshot: bool, cp1251: bool) {
    let dir = dir.as_ref();
    let dir_display = dir.display();
    let dir_content = fs::read_dir(dir).unwrap_or_else(|e| {
        panic!(
            "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
        )
    });

    eprintln!("Testing dbc files in directory: {dir_display}");
    for dbc_path in dir_content {
        let path = dbc_path.unwrap().path();
        if path
            .extension()
            .is_some_and(|ext| ext.eq_ignore_ascii_case("dbc"))
        {
            parse_one_file(path, snapshot, cp1251);
        }
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file(path: impl AsRef<Path>, snapshot: bool, cp1251: bool) {
    let path = path.as_ref();
    eprintln!("Testing DBC file: {}", path.display());
    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    let buffer = if cp1251 {
        decode_cp1252(&buffer)
            .unwrap_or_else(|| panic!("Failed to decode {} as cp1252", path.display()))
    } else {
        Cow::Borrowed(
            std::str::from_utf8(&buffer)
                .unwrap_or_else(|_| panic!("Failed to decode {} as utf-8", path.display())),
        )
    };

    match DbcParser::parse(Rule::file, &buffer) {
        Ok(pairs) => {
            if std::env::var("SKIP_INSTA").is_err()
                && (snapshot || std::env::var("FORCE_INSTA").is_ok())
            {
                assert_debug_snapshot!(file_name, pairs);
            }
        }
        Err(e) => panic!("Failed to parse {file_name}.dbc: {e:#?}"),
    }
}
