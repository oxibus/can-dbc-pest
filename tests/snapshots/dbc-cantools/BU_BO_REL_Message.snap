---
source: tests/snapshots.rs
---
[
    Pair {
        rule: file,
        span: Span {
            str: "VERSION \"\"\n\n\nNS_ : \n\tNS_DESC_\n\tCM_\n\tBA_DEF_\n\tBA_\n\tVAL_\n\tCAT_DEF_\n\tCAT_\n\tFILTER\n\tBA_DEF_DEF_\n\tEV_DATA_\n\tENVVAR_DATA_\n\tSGTYPE_\n\tSGTYPE_VAL_\n\tBA_DEF_SGTYPE_\n\tBA_SGTYPE_\n\tSIG_TYPE_REF_\n\tVAL_TABLE_\n\tSIG_GROUP_\n\tSIG_VALTYPE_\n\tSIGTYPE_VALTYPE_\n\tBO_TX_BU_\n\tBA_DEF_REL_\n\tBA_REL_\n\tBA_DEF_DEF_REL_\n\tBU_SG_REL_\n\tBU_EV_REL_\n\tBU_BO_REL_\n\tSG_MUL_VAL_\n\nBS_:\n\nBU_: ECU1 ECU2\n\n\nBO_ 1 message_1: 7 ECU1\n SG_ signal_1 : 0|8@1+ (1,0) [0|100] \"\"  ECU2\n\n\n\n\n\nBA_DEF_ BO_  \"GenMsgCycleTime\" INT 0 65535;\nBA_DEF_REL_ BU_BO_REL_  \"MsgProject\" ENUM  \"A\",\"B\",\"C\";\nBA_DEF_DEF_  \"GenMsgCycleTime\" 0;\nBA_DEF_DEF_REL_ \"MsgProject\" \"A\";\n\nBA_REL_ \"MsgProject\" BU_BO_REL_ ECU1 1 2;\nVAL_ 1 signal_1 255 \"Invalid\" ;\n\n\n\n",
            start: 0,
            end: 681,
        },
        inner: [
            Pair {
                rule: version,
                span: Span {
                    str: "VERSION \"\"\n",
                    start: 0,
                    end: 11,
                },
                inner: [
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"\"",
                            start: 8,
                            end: 10,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "",
                                    start: 9,
                                    end: 9,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: new_symbols,
                span: Span {
                    str: "NS_ : \n\tNS_DESC_\n\tCM_\n\tBA_DEF_\n\tBA_\n\tVAL_\n\tCAT_DEF_\n\tCAT_\n\tFILTER\n\tBA_DEF_DEF_\n\tEV_DATA_\n\tENVVAR_DATA_\n\tSGTYPE_\n\tSGTYPE_VAL_\n\tBA_DEF_SGTYPE_\n\tBA_SGTYPE_\n\tSIG_TYPE_REF_\n\tVAL_TABLE_\n\tSIG_GROUP_\n\tSIG_VALTYPE_\n\tSIGTYPE_VALTYPE_\n\tBO_TX_BU_\n\tBA_DEF_REL_\n\tBA_REL_\n\tBA_DEF_DEF_REL_\n\tBU_SG_REL_\n\tBU_EV_REL_\n\tBU_BO_REL_\n\tSG_MUL_VAL_\n\n",
                    start: 13,
                    end: 337,
                },
                inner: [
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "NS_DESC_",
                            start: 21,
                            end: 29,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "CM_",
                            start: 31,
                            end: 34,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_",
                            start: 36,
                            end: 43,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_",
                            start: 45,
                            end: 48,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "VAL_",
                            start: 50,
                            end: 54,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "CAT_DEF_",
                            start: 56,
                            end: 64,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "CAT_",
                            start: 66,
                            end: 70,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "FILTER",
                            start: 72,
                            end: 78,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_DEF_",
                            start: 80,
                            end: 91,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "EV_DATA_",
                            start: 93,
                            end: 101,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "ENVVAR_DATA_",
                            start: 103,
                            end: 115,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SGTYPE_",
                            start: 117,
                            end: 124,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SGTYPE_VAL_",
                            start: 126,
                            end: 137,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_SGTYPE_",
                            start: 139,
                            end: 153,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_SGTYPE_",
                            start: 155,
                            end: 165,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIG_TYPE_REF_",
                            start: 167,
                            end: 180,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "VAL_TABLE_",
                            start: 182,
                            end: 192,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIG_GROUP_",
                            start: 194,
                            end: 204,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIG_VALTYPE_",
                            start: 206,
                            end: 218,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIGTYPE_VALTYPE_",
                            start: 220,
                            end: 236,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BO_TX_BU_",
                            start: 238,
                            end: 247,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_REL_",
                            start: 249,
                            end: 260,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_REL_",
                            start: 262,
                            end: 269,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_DEF_REL_",
                            start: 271,
                            end: 286,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BU_SG_REL_",
                            start: 288,
                            end: 298,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BU_EV_REL_",
                            start: 300,
                            end: 310,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BU_BO_REL_",
                            start: 312,
                            end: 322,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SG_MUL_VAL_",
                            start: 324,
                            end: 335,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: bit_timing,
                span: Span {
                    str: "BS_:\n",
                    start: 337,
                    end: 342,
                },
                inner: [],
            },
            Pair {
                rule: nodes,
                span: Span {
                    str: "BU_: ECU1 ECU2\n\n\n",
                    start: 343,
                    end: 360,
                },
                inner: [
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "ECU1",
                            start: 348,
                            end: 352,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "ECU2",
                            start: 353,
                            end: 357,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: message,
                span: Span {
                    str: "BO_ 1 message_1: 7 ECU1",
                    start: 360,
                    end: 383,
                },
                inner: [
                    Pair {
                        rule: can_id,
                        span: Span {
                            str: "1",
                            start: 364,
                            end: 365,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident_ext,
                        span: Span {
                            str: "message_1",
                            start: 366,
                            end: 375,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "7",
                            start: 377,
                            end: 378,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "ECU1",
                            start: 379,
                            end: 383,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: signal,
                span: Span {
                    str: "SG_ signal_1 : 0|8@1+ (1,0) [0|100] \"\"  ECU2",
                    start: 385,
                    end: 429,
                },
                inner: [
                    Pair {
                        rule: ident_ext,
                        span: Span {
                            str: "signal_1",
                            start: 389,
                            end: 397,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "0",
                            start: 400,
                            end: 401,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "8",
                            start: 402,
                            end: 403,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "1",
                            start: 404,
                            end: 405,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "1",
                            start: 408,
                            end: 409,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0",
                            start: 410,
                            end: 411,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0",
                            start: 414,
                            end: 415,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "100",
                            start: 416,
                            end: 419,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"\"",
                            start: 421,
                            end: 423,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "",
                                    start: 422,
                                    end: 422,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "ECU2",
                            start: 425,
                            end: 429,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: attr_def,
                span: Span {
                    str: "BA_DEF_ BO_  \"GenMsgCycleTime\" INT 0 65535;",
                    start: 435,
                    end: 478,
                },
                inner: [
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"GenMsgCycleTime\"",
                            start: 448,
                            end: 465,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "GenMsgCycleTime",
                                    start: 449,
                                    end: 464,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0",
                            start: 470,
                            end: 471,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "65535",
                            start: 472,
                            end: 477,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: ba_def_rel,
                span: Span {
                    str: "BA_DEF_REL_ BU_BO_REL_  \"MsgProject\" ENUM  \"A\",\"B\",\"C\";",
                    start: 479,
                    end: 534,
                },
                inner: [
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"MsgProject\"",
                            start: 503,
                            end: 515,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "MsgProject",
                                    start: 504,
                                    end: 514,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"A\"",
                            start: 522,
                            end: 525,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "A",
                                    start: 523,
                                    end: 524,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"B\"",
                            start: 526,
                            end: 529,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "B",
                                    start: 527,
                                    end: 528,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"C\"",
                            start: 530,
                            end: 533,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "C",
                                    start: 531,
                                    end: 532,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: ba_def_def,
                span: Span {
                    str: "BA_DEF_DEF_  \"GenMsgCycleTime\" 0;",
                    start: 535,
                    end: 568,
                },
                inner: [
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"GenMsgCycleTime\"",
                            start: 548,
                            end: 565,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "GenMsgCycleTime",
                                    start: 549,
                                    end: 564,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0",
                            start: 566,
                            end: 567,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: ba_def_def_rel,
                span: Span {
                    str: "BA_DEF_DEF_REL_ \"MsgProject\" \"A\";",
                    start: 569,
                    end: 602,
                },
                inner: [
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"MsgProject\"",
                            start: 585,
                            end: 597,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "MsgProject",
                                    start: 586,
                                    end: 596,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"A\"",
                            start: 598,
                            end: 601,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "A",
                                    start: 599,
                                    end: 600,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: ba_rel,
                span: Span {
                    str: "BA_REL_ \"MsgProject\" BU_BO_REL_ ECU1 1 2;",
                    start: 604,
                    end: 645,
                },
                inner: [
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"MsgProject\"",
                            start: 612,
                            end: 624,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "MsgProject",
                                    start: 613,
                                    end: 623,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "ECU1",
                            start: 636,
                            end: 640,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "1",
                            start: 641,
                            end: 642,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "2",
                            start: 643,
                            end: 644,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: value_table_def,
                span: Span {
                    str: "VAL_ 1 signal_1 255 \"Invalid\" ;",
                    start: 646,
                    end: 677,
                },
                inner: [
                    Pair {
                        rule: can_id,
                        span: Span {
                            str: "1",
                            start: 651,
                            end: 652,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "signal_1",
                            start: 653,
                            end: 661,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "255",
                            start: 662,
                            end: 665,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"Invalid\"",
                            start: 666,
                            end: 675,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "Invalid",
                                    start: 667,
                                    end: 674,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: EOI,
                span: Span {
                    str: "",
                    start: 681,
                    end: 681,
                },
                inner: [],
            },
        ],
    },
]
